slim_build_context.md

```diff
@@ -0,0 +1,162 @@
+# Slim Build Context (Proposal)
+
+## Problem
+Generally, Tilt does not give users great control over what kinds of changes
+cause builds to happen. Tilt (reasonably) aims to ensure that the user is
+never running incorrect or out-of-date code, and the current implementation
+leads to the user often waiting for builds and deploys that they really don't
+need.
+
+e.g.:
+1. In [tiltdemo](https://github.com/windmilleng/tiltdemo/blob/f4a6aa3d0c43357ce0bb39bbc4ff5f37cfd2ffb8/Tiltfile),
+we have two services, which each have '.' as the build context. If you change
+demoserver1, you get a live update of demoserver1 and a full image build of
+demoserver2, and vice versa. In order to prevent this, we'd have to dramatically
+restructure the repo.
+2. In tilt.build, Nick tried the [intuitive](https://github.com/windmilleng/tilt.build/commit/bd251498ceffae56f0493eb772f4a3399517f406) option to switching to `live_update`,
+and this led to `tilt-site` doing a full build when files in `docs` changed,
+and he had to restructure things to get around that.
+3. Images with very simple base images, e.g.:
+```
+docker_build('base', '.', dockerfile='base.dockerfile')
+img = docker_build('app', '.')
+img.add('.', '/app')
+```
+(with `app`'s Dockerfile containing a `FROM base`).
+It is perfecctly reasonable for a user to define their config like this, with any changes
+to `app` only relevant to the `app` image. However, since the build context for
+the base image is `.`, any changes to `app` will invalidate `base`, causing
+a full image build for `base`, `app`, and any other images/resources that depend
+on `base`. It's effectively impossible to configure live updates for `app`
+without moving `base` into a subdirectory, which is not very "meet the customer
+where they are".
+
+## Non-goals
+Reducing how much data is sent to the docker daemon. While this is often excessive,
+it's a thing people are used to, because it's how docker works, and it's not clearly
+a source of pain.
+
+## Solution Options
+Brief descriptions here, pros/cons in the next section.
+1. Watch only an image's actual dependencies instead of its full build context.
+  1. e.g., parse each Dockerfile, find ADDs and COPYs, and only watch the files
+     that get added and copied.
+2. Add a `docker_build` option `only_rebuild_on` - only files listed here will
+   trigger builds, but the build still has access to the full build context, e.g.:
+   ```
+   docker_build('base', '.', dockerfile='base.dockerfile', only_rebuild_on=['package.json'])
+   ```
+3. Add `docker_build` options `ignore` and/or `only`, which filter the build
+   context - both the files that are watched and the files that are included
+   for the build, e.g.:
+   ```
+   docker_build('base', '.', dockerfile='base.dockerfile', ignore=['**', '!package.json'])
+   ```
+4. Allow `docker_build`'s context param to take either the string (directory)
```

Comment:
I don't see pros/cons for this, which is fine. I'm guessing it's about how it's more work and another layer of object/config?

Reply:
pros/cons are in the next section. is that what you're looking for?

![image](https://user-images.githubusercontent.com/7453991/56389744-53f72400-61f8-11e9-9eb0-ee67eef30602.png)

(though there's an extra newline there that I need to fix)